proto/cacheclient.proto

```diff
@@ -17,36 +17,90 @@ enum ECacheResult {
 }
 
 service Scs {
-  rpc Get (_GetRequest) returns (_GetResponse) {}
-  rpc Set (_SetRequest) returns (_SetResponse) {}
-  rpc SetIfNotExists (_SetIfNotExistsRequest) returns (_SetIfNotExistsResponse) {}
-  rpc Delete (_DeleteRequest) returns (_DeleteResponse) {}
-  rpc KeysExist (_KeysExistRequest) returns (_KeysExistResponse) {}
-  rpc Increment (_IncrementRequest) returns (_IncrementResponse) {}
-  rpc UpdateTtl (_UpdateTtlRequest) returns (_UpdateTtlResponse) {}
-
-  rpc DictionaryGet (_DictionaryGetRequest) returns (_DictionaryGetResponse) {}
-  rpc DictionaryFetch (_DictionaryFetchRequest) returns (_DictionaryFetchResponse) {}
-  rpc DictionarySet (_DictionarySetRequest) returns (_DictionarySetResponse) {}
-  rpc DictionaryIncrement (_DictionaryIncrementRequest) returns (_DictionaryIncrementResponse) {}
-  rpc DictionaryDelete (_DictionaryDeleteRequest) returns (_DictionaryDeleteResponse) {}
-
-  rpc SetFetch (_SetFetchRequest) returns (_SetFetchResponse) {}
-  rpc SetUnion (_SetUnionRequest) returns (_SetUnionResponse) {}
-  rpc SetDifference (_SetDifferenceRequest) returns (_SetDifferenceResponse) {}
-  rpc SetContains (_SetContainsRequest) returns (_SetContainsResponse) {}
-
-  rpc ListPushFront(_ListPushFrontRequest) returns (_ListPushFrontResponse) {}
-  rpc ListPushBack(_ListPushBackRequest) returns (_ListPushBackResponse) {}
-  rpc ListPopFront(_ListPopFrontRequest) returns (_ListPopFrontResponse) {}
-  rpc ListPopBack(_ListPopBackRequest) returns (_ListPopBackResponse) {}
-  rpc ListErase(_ListEraseRequest) returns (_ListEraseResponse) {}
-  rpc ListRemove(_ListRemoveRequest) returns (_ListRemoveResponse) {}
-  rpc ListFetch(_ListFetchRequest) returns (_ListFetchResponse) {}
-  rpc ListLength(_ListLengthRequest) returns (_ListLengthResponse) {}
-  rpc ListConcatenateFront(_ListConcatenateFrontRequest) returns (_ListConcatenateFrontResponse) {}
-  rpc ListConcatenateBack(_ListConcatenateBackRequest) returns (_ListConcatenateBackResponse) {}
-  rpc ListRetain(_ListRetainRequest) returns (_ListRetainResponse) {}
+  rpc Get (_GetRequest) returns (_GetResponse) {
+    option idempotency_level = NO_SIDE_EFFECTS;
+  }
+  rpc Set (_SetRequest) returns (_SetResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
+  rpc SetIfNotExists (_SetIfNotExistsRequest) returns (_SetIfNotExistsResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
+  rpc Delete (_DeleteRequest) returns (_DeleteResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
+  rpc KeysExist (_KeysExistRequest) returns (_KeysExistResponse) {
+    option idempotency_level = NO_SIDE_EFFECTS;
+  }
+  rpc Increment (_IncrementRequest) returns (_IncrementResponse) {
+    option idempotency_level = IDEMPOTENCY_UNKNOWN;
+  }
+  rpc UpdateTtl (_UpdateTtlRequest) returns (_UpdateTtlResponse) {
+    option idempotency_level = IDEMPOTENT;
+  }
```

Comment:
this is not idempotent. Look at the update ttl request message

Reply:
I was wondering about this one. I assumed idempotent because I couldn't think of a case where a first try would not update the ttl but an additional try would. Setting it to the same ttl multiple times wouldn't be a problem.

Reply:
> Setting it to the same ttl multiple times wouldn't be a problem.

Can you explain the result of each of these if re-executed? https://github.com/momentohq/client-protos/blob/main/proto/cacheclient.proto#L146-L153 

fyi "wouldn't be a problem" is not a sufficient condition for an action to be correctly described as "idempotent." A quick google should help clear the meaning up!

Reply:
An idempotent function is one that has no further effect after the first call, or rather can be called any number of times with the same effect as being called one time. I would appreciate if you told me the specific problem you have instead of asking me to google word definitions.

If an object has a ttl of 1000 and I call this function to set it to 2000:
increase_to_milliseconds: The first call sets it to 2000 and subsequent calls leave it at 2000.
decrease_to_milliseconds: The first call leaves it at 1000 as do subsequent calls
overwrite_to_milliseconds: The first call sets it to 2000 and subsequent calls leave it at 2000.

If an object has a ttl of 3000 and I set it to 2000:
increase_to_milliseconds: The first call leaves the ttl at 3000 as do subsequent calls.
decrease_to_milliseconds: The first call sets it to 2000 and subsequent calls leave it at 2000.
overwrite_to_milliseconds: The first call sets it to 2000 and subsequent calls leave it at 2000.

I certainly could be missing a case or not considering side effects like the ttl resetting, but I don't immediately see it. What is your specific objection? How is setting a ttl to a specific number 5 times any different from setting it once?

Reply:
> An idempotent function is one that has no further effect after the first call, or rather can be called any number of times with the same effect as being called one time. I would appreciate if you told me the specific problem you have instead of asking me to google word definitions.

Cool, sounds like you understand the word now. Thanks. It's important to understand what you are PRing. In response to your normative request, I am unable to vend specific answers for nonspecific misunderstandings, misrepresentations or in the presence of material ambiguity. Though I am sympathetic toward that which I gather you might appreciate, I will be broadly unable to accommodate my understanding of the request now or in the future on technical pull requests afflicted by any the aforementioned.

Thanks for the clear listing. I'll address or correct each inline (leaving some inaccurate phrasing in place to focus on the more salient aspects):

> If an object has ~a ttl of 1000~ 1 second of ttl remaining and I call this function to set it to 2000 milliseconds:

* increase_to_milliseconds: The first call sets it to ~2000~ $(( `date -u +%s` + 2 )) and subsequent calls ~leave it at 2000~ set it to $(( `date -u +%s` + 2 )).
* decrease_to_milliseconds: The first call leaves it at 1000 as do subsequent calls
* overwrite_to_milliseconds: The first call sets it to ~2000~ $(( `date -u +%s` + 2 )) and subsequent calls ~leave it at 2000~ set it to $(( `date -u +%s` + 2 )).

> If an object has ~a ttl of 3000~ 3 seconds of ttl remaining and I set it to 2000:

* increase_to_milliseconds: The first call leaves the ttl at 3000 ~as do subsequent calls.~ and subsequent calls, if they come at least 1 second later, update the ttl to $(( `date -u +%s` + 2 )).
* decrease_to_milliseconds: The first call sets it to 2000 and subsequent calls leave it at 2000.
* overwrite_to_milliseconds: The first call sets it to 2000 and subsequent calls ~leave it at 2000.~ update the ttl to $(( `date -u +%s` + 2 )).

I hope these clarifications help you identify the mental model disconnect, but if there is any additional clarification I can provide feel free to reach out.

Reply:
I'll update the pr. The docs on the definition for that function don't specify whether it refreshes the ttl, and I assumed it did not. In the future, feel free to simply state why my assumption is incorrect.

Reply:
> The docs on the definition for that function don't specify whether it refreshes the ttl, and I assumed it did not.

I'm not sure what you mean by this and it might indicate a continued misunderstanding. In your examples you seem to (correctly) assume the ttl's in the request object represent relative ttl values. The dimension of time, however, does not pause between api invocations, so a relative duration value would be interpreted differently based on the context of different arrival times.

> In the future, feel free to simply state why my assumption is incorrect.

As I tried to clarify before, I will be unable to do this now or in the future _particularly_ when your assumptions are not made clear. This conversation is no longer relevant to this PR. We can follow up on slack if you're interested in pursuing it further.